### ISSUE TITLE
[Idea] A RenderTarget widget that can capture mouse and report whether it is active

### ISSUE THREAD

**Issue by Clocktown** [ocornut/imgui#2169]
**Version/Branch of Dear ImGui:**
Latest master/docking/viewport

**Back-end file/Renderer/OS:**
All Backends that support mouse capture, GLFW example code provided.
OS: All OS, example Code tested on Windows

**My Idea:** 
***Preliminary sweet-talk***
Up to now my usage of ImGui was mostly rendering to the full main window and having some floating ImGui-Windows flying around for configuration, maybe with a main menu. You'd just render to the default framebuffer and that's that. Proper Input/Callback handling was easily reached with checks for `io.WantCaptureKeyboard` and `io.WantCaptureMouse` and the like.  
Multiple Viewports make input handling a bit more complicated, since we now have several window handles and all of the newly created windows are automatically wired up to the callbacks provided in the backend file.  
Then we also have docking. This awesome feature greatly expands the usage of ImGui. We can now build Apps with a nice GUI-Layout, e.g. using a DockSpace. With this, however, comes the need to Render into Textures and display these in Windows instead of just targeting the area behind our floating windows. This is easily possible with `ImGui::Image`, however, properly handling input can now become a problem. What you'd typically want is to use different callbacks/input logic depending on whether the GUI is used or whether the mouse is inside that Image being rendered to, e.g. for camera controls which should not happen when you are modifying a slider. Since the Image is part of ImGui, it would still report to want to capture the mouse. Combine this with multiple viewports and maybe different windows with different images that should use different Input logic and chaos is ensured.  

***Interesting part***
I propose to introduce a new wrapper widget for `ImGui::Image` which addresses these issues, as well as some additional functions that have to be wired up to the window manager backend and the like.  
I provided a working example of such a widget below. The example also allows to capture the mouse cursor which is a very nice feature to have for camera controls and the like. This is triggered by right-clicking, but it would be easy to add configuration options for this.  
It currently has GLFW logic hard-wired, ideally we'd want to replace this with functions `ImGui::ToggleMouseCaptured()` and ` ImGui::IsMouseCaptured()` which are wired up to the respective backend. For GLFW, both the main window as well as the actual window of the viewport have to be changed.  
The widget will report whether it is active in a boolean (it is active if it is being hovered or if it caused the mouse to be captured previously). This boolean - alongside with information whether the surrounding window is open/closed/collapsed - can then be used by the user to ensure the correct callbacks and input logic happening.  
One thing that is quite important is to change the mouse position update logic inside the backend code. The mouse position should not be updated for ImGui if the mouse is captured - otherwise the behaviour will be very strange.  

There is one _major problem_ with the provided implementation. If the mouse is captured and subsequently the `RenderTarget` window is moved away from its original position or something similiar, such that the mouse position before capturing is not hovering the widget anymore, the mouse cannot be freed again.

**Example Code** 
```c++
IMGUI_API void RenderTarget(ImTextureID user_texture_id, const ImVec2& size, bool* p_active,
                            const ImVec2& uv0 = ImVec2(0, 0), const ImVec2& uv1 = ImVec2(1, 1),
                            const ImVec4& tint_col   = ImVec4(1, 1, 1, 1),
                            const ImVec4& border_col = ImVec4(0, 0, 0, 0))
{
    ImGui::Image(user_texture_id, size, uv0, uv1, tint_col, border_col);
    ImGuiMouseCursor imgui_cursor = ImGui::GetMouseCursor();
    auto             mwind        = (GLFWwindow*)ImGui::GetMainViewport()->PlatformHandle;
    auto             wind         = (GLFWwindow*)ImGui::GetWindowViewport()->PlatformHandle;
    auto             mode         = glfwGetInputMode(mwind, GLFW_CURSOR);
    if(ImGui::IsItemClicked(1) && mode != GLFW_CURSOR_DISABLED)
    {
        glfwSetInputMode(mwind, GLFW_CURSOR, GLFW_CURSOR_DISABLED);
        glfwSetInputMode(wind, GLFW_CURSOR, GLFW_CURSOR_DISABLED);
    }

    if(ImGui::IsItemClicked(1) && mode == GLFW_CURSOR_DISABLED)
    {
        if(imgui_cursor == ImGuiMouseCursor_None || ImGui::GetIO().MouseDrawCursor)
        {
            // Hide OS mouse cursor if imgui is drawing it or if it wants no cursor
            glfwSetInputMode(mwind, GLFW_CURSOR, GLFW_CURSOR_HIDDEN);
            glfwSetInputMode(wind, GLFW_CURSOR, GLFW_CURSOR_HIDDEN);
        }
        else
        {
            glfwSetInputMode(mwind, GLFW_CURSOR, GLFW_CURSOR_NORMAL);
            glfwSetInputMode(wind, GLFW_CURSOR, GLFW_CURSOR_NORMAL);
        }
    }
    if(ImGui::IsItemHovered())
    {
        *p_active = true;
    }
    else
    {
        *p_active = false;
    }
}
```
Changes to the backend mouse position update are also quite simple, e.g. for GLFW inside `ImGui_ImplGlfw_UpdateMousePosAndButtons`:
```c++
[... snip ...]
        if(focused)
        {
            // IMPORTANT - For proper 3D Viewports with camera controls etc, we want to capture the
            // mouse sometimes using GLFW If mouse is captured, prevent ImGui from being updated
            // accordingly. Otherwise the invisible mouse will be able to interact with widgets!
            // It is enough to check the main viewport, because this always has to be set
            // accordingly in order for mouse capture to work.
            if(glfwGetInputMode((GLFWwindow*)ImGui::GetMainViewport()->PlatformHandle,
                                GLFW_CURSOR) != GLFW_CURSOR_DISABLED)
            {
                if(io.WantSetMousePos)
                {
                    glfwSetCursorPos(window,
                                     (double)(mouse_pos_backup.x - viewport->Pos.x),
                                     (double)(mouse_pos_backup.y - viewport->Pos.y));
                }
                else
                {
                    double mouse_x, mouse_y;
                    glfwGetCursorPos(window, &mouse_x, &mouse_y);
                    io.MousePos = ImVec2((float)mouse_x + viewport->Pos.x,
                                         (float)mouse_y + viewport->Pos.y);
                }
            }
            else
            {
                // Restore old position, such that the image that triggered the capture will still
                // report hovered
                io.MousePos = mouse_pos_backup;
            }
            for(int i = 0; i < IM_ARRAYSIZE(io.MouseDown); i++)
                io.MouseDown[i] |= glfwGetMouseButton(window, i) != 0;
        }
[... snip ...]
```

**Comment by Clocktown** [comment]
The remaining issue with my implementation I mentioned was bothering me a lot. I came up with this solution:
```c++
// Handle free mouse
        auto& io = ImGui::GetIO();
        if(ImGui::IsMouseClicked(1))
        {
            if(glfwGetInputMode((GLFWwindow*)ImGui::GetMainViewport()->PlatformHandle,
                                GLFW_CURSOR) == GLFW_CURSOR_DISABLED)
            {
                io.MouseDownDuration[1]       = 0.01f;
                ImGuiPlatformIO& platform_io  = ImGui::GetPlatformIO();
                ImGuiMouseCursor imgui_cursor = ImGui::GetMouseCursor();
                for(int n = 0; n < platform_io.Viewports.Size; n++)
                {
                    ImGuiViewport* viewport = platform_io.Viewports[n];
                    GLFWwindow*    window   = (GLFWwindow*)viewport->PlatformHandle;
                    if(imgui_cursor == ImGuiMouseCursor_None || ImGui::GetIO().MouseDrawCursor)
                    {
                        // Hide OS mouse cursor if imgui is drawing it or if it wants no cursor
                        glfwSetInputMode(window, GLFW_CURSOR, GLFW_CURSOR_HIDDEN);
                    }
                    else
                    {
                        glfwSetInputMode(window, GLFW_CURSOR, GLFW_CURSOR_NORMAL);
                    }
                }
            }
        }
```

This has to be called after `ImGui::NewFrame`, or rather, after `ImGui::NewFrame` has updated mouse button state. This handles freeing of the mouse at a right-click no matter what, even if one manages to move away the window from under the captured cursor.  
The widget implementation then shortens accordingly:  
```c++
IMGUI_API void RenderTarget(ImTextureID user_texture_id, const ImVec2& size, bool* p_active,
                            const ImVec2& uv0 = ImVec2(0, 0), const ImVec2& uv1 = ImVec2(1, 1),
                            const ImVec4& tint_col   = ImVec4(1, 1, 1, 1),
                            const ImVec4& border_col = ImVec4(0, 0, 0, 0))
{
    ImGui::Image(user_texture_id, size, uv0, uv1, tint_col, border_col);
    ImGuiMouseCursor imgui_cursor = ImGui::GetMouseCursor();
    auto             mwind        = (GLFWwindow*)ImGui::GetMainViewport()->PlatformHandle;
    auto             wind         = (GLFWwindow*)ImGui::GetWindowViewport()->PlatformHandle;
    auto             mode         = glfwGetInputMode(mwind, GLFW_CURSOR);
    if(ImGui::IsItemClicked(1) && mode != GLFW_CURSOR_DISABLED)
    {
        glfwSetInputMode(mwind, GLFW_CURSOR, GLFW_CURSOR_DISABLED);
        glfwSetInputMode(wind, GLFW_CURSOR, GLFW_CURSOR_DISABLED);
    }

    if(ImGui::IsItemHovered())
    {
        *p_active = true;
    }
    else
    {
        *p_active = false;
    }
}
```

Again: All of this needs to be made backend agnostic still.

**Comment by ocornut** [comment]
Hello,

I understand the issue you describe but I think it would be best solved on the user's side. 
Your suggestion is adding too much contact surface with the underlying input system while being restrictive on what can be done. 

> // IMPORTANT - For proper 3D Viewports with camera controls etc, we want to capture the
 > // mouse sometimes using GLFW If mouse is captured, prevent ImGui from being updated
 > // accordingly. Otherwise the invisible mouse will be able to interact with widgets!

You can set `ImGuiConfigFlags_NoMouse` for this purpose.

I think all that has to be done is:
- Have your window/widget set a global boolean flag e.g. "bool IsControllingGame"
- When set, your code can 1) hide the mouse cursor using `ImGui::SetMouseCursor(ImGuiMouseCursor_None)` or your own mean, 2) set `ImGuiConfigFlags_NoMouse` so imgui doesn't see the mouse, and 3) call the os/framework to disable your mouse interactions globally e.g. using GLFW's GLFW_CURSOR_HIDDEN.
- You'll need to provide a way to get out of this mode (e.g. using the ESC key).

Consider this example (which I made a while ago but not added to imgui_demo.cpp because unfinished):
(The example as-is is weird because the actual ball controls is using the keyboard not the mouse)

![inputs](https://user-images.githubusercontent.com/8225057/48002959-5c0d4c80-e10d-11e8-85e1-a790be6080ec.gif)

```
static void ShowExampleAppGameCaptureInputs(bool* p_open)
{
    if (!ImGui::Begin("Example: Game capture inputs", p_open, ImGuiWindowFlags_AlwaysAutoResize))
    {
        ImGui::End();
        return;
    }

    ImGuiIO& io = ImGui::GetIO();
    ImDrawList* draw_list = ImGui::GetWindowDrawList();
    ImVec2 sz = ImVec2(1920 / 4.0f, 1080 / 4.0f);

    static bool game_has_inputs = false;
    static ImVec2 hero_pos = ImVec2(100.0f, 0.0f);
    static ImVec2 hero_vel = ImVec2(0.0f, 0.0f);
    static float physics_time_accum = 0.0f;

    // Update Game Inputs
    if (game_has_inputs)
    {
        // Request imgui to ignore mouse. 
        // Request back-end to hide the mouse cursor (NB: back-end will ignore this if ImGuiConfigFlags_NoMouseCursorChange is enabled)
        // THIS IS INCOMPLETE AND SHORT-SIGHTED! In your own game/application, you probably want to capture the cursor, keep it in a same OS position etc.
        io.ConfigFlags |= ImGuiConfigFlags_NoMouse;
        ImGui::SetMouseCursor(ImGuiMouseCursor_None);

        // For the scope of this demo we are reading inputs via imgui helpers, which is a rather odd and misleading, but this is all we have here.
        // In your own game/application you will likely want use your own input systems, and it would be perfectly fine to decide to stop providing any
        // inputs to imgui when your game capture.
        if (ImGui::IsKeyDown(ImGui::GetKeyIndex(ImGuiKey_LeftArrow)))  hero_vel.x -= 4000.0f * io.DeltaTime;
        if (ImGui::IsKeyDown(ImGui::GetKeyIndex(ImGuiKey_RightArrow))) hero_vel.x += 4000.0f * io.DeltaTime;
        if (ImGui::IsKeyDown(ImGui::GetKeyIndex(ImGuiKey_UpArrow)))    hero_vel.y += 4000.0f * io.DeltaTime;
        if (ImGui::IsKeyDown(ImGui::GetKeyIndex(ImGuiKey_Escape))) 
        { 
            game_has_inputs = false; 
            io.ConfigFlags &= ~ImGuiConfigFlags_NoMouse; 
        }
    }

    // Update Dumb Physics
    float TIME_STEP = 1.0f / 120.0f;
    physics_time_accum += io.DeltaTime;
    while (physics_time_accum >= TIME_STEP)
    {
        hero_pos.x += hero_vel.x * TIME_STEP;                                  // Integrate position
        hero_pos.y += hero_vel.y * TIME_STEP;                                 // Integrate position
        hero_vel.x = fabsf(hero_vel.x) < 0.0001f ? 0.0f : hero_vel.x * 0.98f; // Dampen velocity
        hero_vel.y = hero_vel.y - 1000.0f * TIME_STEP;
        if (hero_pos.x < 0.0f) { hero_pos.x = 0.0f; hero_vel.x *= -0.90f; }      // Bounce 
        if (hero_pos.x > sz.x) { hero_pos.x = sz.x; hero_vel.x *= -0.90f; }      // Bounce
        if (hero_pos.y < 40.f) { hero_pos.y = 40.f; hero_vel.y = fabsf(hero_vel.y) < 50.0f ? 0.0f : hero_vel.y *= -0.70f; }
        if (hero_pos.y > sz.y) { hero_pos.y = sz.y; hero_vel.y *= -0.70f; }
        physics_time_accum -= TIME_STEP;
    }

    // Draw game inside the window
    // In a real usage scenario, our game/application will probably have been rendered to a texture, so you can display it with a simple call 
    //  to e.g. ImGui::Image(my_framebuffer, size, ImVec2(0,0), ImVec2(1,1));
    // This demo code being platform/engine dependent we cannot render to a texture so we draw the output programmatically.
    {
        ImVec2 p1 = ImGui::GetCursorScreenPos();
        ImVec2 p2 = ImVec2(p1.x + sz.x, p1.y + sz.y);
        draw_list->PushClipRect(p1, p2, true);
        const float radius = 32.0f;
        const ImVec2 fwd = ImVec2(cosf(hero_pos.x / radius) * radius, sinf(hero_pos.x / radius) * radius);          // Offset from center to front point according to current angle
        const ImVec2 pos = ImVec2(p1.x + hero_pos.x, p1.y - hero_pos.y + sz.y);                                     // Screen position
        draw_list->AddRectFilled(p1, p2, IM_COL32(161, 173, 255, 255));                                             // Background
        draw_list->AddRectFilled(ImVec2(p1.x, p1.y + sz.y * 0.9f), ImVec2(p2.x, p2.y), IM_COL32(222, 89, 24, 255)); // Floor
        draw_list->AddCircleFilled(pos, radius, IM_COL32(255, 180, 60, 255), 16);                                   // Hero
        draw_list->AddLine(ImVec2(pos.x - fwd.x, pos.y - fwd.y), ImVec2(pos.x + fwd.x, pos.y + fwd.y), IM_COL32(0, 0, 0, 255), 2.0f); // Hero Line
        draw_list->PopClipRect();
        if (game_has_inputs)
            draw_list->AddRect(p1, p2, IM_COL32(255, 255, 255, 255));
    }
    if (ImGui::InvisibleButton("##game", sz))
        game_has_inputs = true;

    ImGui::Text("game_has_inputs = %d%s", game_has_inputs, game_has_inputs ? " (Press escape to return to UI)" : "");
    ImGui::SameLine(); ShowHelpMarker("Please read the code/comments carefully to understand what this demo is doing.");
    ImGui::Text("pos (%.1f,%.1f) vel (%.1f,%.1f)", hero_pos.x, hero_pos.y, hero_vel.x, hero_vel.y);

    ImGui::End();
}
```

**Comment by Clocktown** [comment]
You're right and it can be done like this. It avoids having to modify the actual backend code like I did, although results are slightly different (as ImGui has mouse completely disabled, while my approach will just report the mouse position always at the position it was before capture).  
You can probably close this issue.  

As a side note,
> 3) call the os/framework to disable your mouse interactions globally e.g. using GLFW's GLFW_CURSOR_HIDDEN.

That wasn't what this was about. With capturing the mouse (using `GLFW_CURSOR_DISABLED` what you achieve is that a) the mouse isn't visible and b) you can move the mouse infinitely in any direction without subject to the monitors boundaries, which is crucial for any decent camera controls. Nevertheless, the approach you posted still applies and works.

**Comment by ocornut** [comment]
Sorry I mixed up `GLFW_CURSOR_HIDDEN` and `GLFW_CURSOR_DISABLED` indeed!

> You can probably close this issue.

If there is anything else you think imgui could provide in order to simplify those usage pattern (similar to the `_NoMouse` flag) let me know! I'd be happy to make that sort of thing easier or more obvious if possible. I'll try to finish the example I posted to demonstrate this example but it's not easy to do it.

I'm not excluding the possibility of providing something (or at least by default?) in the back-ends if we end up finding out that a user side implementation is too tricky to implement.
